### DISCUSSION TITLE
Hydrating from Local Storage on First Render

### DISCUSSION THREAD

**Discussion by cveering** [pmndrs/jotai#1737] [Q&A]
Hi,

Looking at the console output for the following example code you can see that the darkModeAtom that was initialized with atomWithStorage prints to the console twice on refresh. The first time is the default value. The second time is the recalled (hydrated) value pulled from local storage. That is fine for the darkmode use case, but it won't work for authentication checking. 
https://codesandbox.io/s/jotai-persistence-vuwi7?from-embed=&file=/src/app.js

I want to store a userAtom in local storage and check for the existence of the user to determine if access to private routes should be granted. 

store.js
```
const userAtom = atomWithStorage('')
```
PrivateRoutesWrapper.jsx
```
const user = useAtom(userAtom)
if (user !== '') { 
    Redirect to login
} else {
    Display private content
}
```

The problem is that on first render, user is always an empty string so it always redirects to login.

Redux with Redux Persist can handle this, but I like Jotai much better. I could always pop the value directly out of local storage, but then all of the other atomic code that relies on user still falls down and things go sideways quickly. Any suggestions for workarounds?

Thanks!

**Comment by cveering** [comment]
So I was able to resolve this by creating a simple atom that grabs the user value directly out of local storage without negatively impacting other functionality. But it would be great if atoms hydrated on first render... I'm only a day into Jotai and am already a big fan.

**Comment by dai-shi** [comment]
The behavior is changed in v2 due to SSR complaining on hydration. Seems like React recommends two renders for SSR.

We could create another util for reading localStorage in sync, but it's rather trivial to create a such atom. And, you have more control explicitly.
https://github.com/pmndrs/jotai/blob/dae4418d330bb8d2ea151fc10ab5bcaaad595bd2/docs/guides/persistence.mdx?plain=1#L14-L22

For AsyncStorage, hm, yeah, it's not that easy.

**Reply by EduardoLopes** [reply]
when i try the example, typescript complains about `newStr` being of type `unknown`. 

I did this:

```js
const strAtom = atom(localStorage.getItem("myKey") ?? "foo");

const strAtomWithPersistence = atom(
  (get) => get(strAtom),
  (get, set, newStr: string) => {
    set(strAtom, newStr);
    localStorage.setItem("myKey", newStr);
  }
);
```

there's a better way?

**Reply by dai-shi** [reply]
that's the best.

**Reply by EduardoLopes** [reply]
This seems like a simpler solution:

```js
const tokenAtom = atomWithStorage<string | undefined>(
  "token",
  localStorage.getItem("token") ?? undefined
);
```

**Reply by dai-shi** [reply]
lol, nice.

**Reply by amirhhashemi** [reply]
Here is a version that uses json storage:
```typescript
import { RESET, unstable_NO_STORAGE_VALUE, createJSONStorage } from "jotai/utils";

function atomWithLocalStorage<Value>(key: string, initialValue: Value) {
  const storage = createJSONStorage<Value>(() =>
    typeof window !== "undefined"
      ? window.localStorage
      : (undefined as unknown as Storage)
  );

  function getInitialValue() {
    const value = storage.getItem(key);
    return value === unstable_NO_STORAGE_VALUE ? initialValue : value;
  }

  const baseAtom = atom(getInitialValue());

  const derivedAtom = atom(
    (get) => get(baseAtom),
    (get, set, update: SetStateActionWithReset<Value>) => {
      const nextValue =
        typeof update === "function"
          ? (update as (prev: Value) => Value | typeof RESET)(get(baseAtom))
          : update;
      if (nextValue === RESET) {
        set(baseAtom, initialValue);
        return storage.removeItem(key);
      }
      set(baseAtom, nextValue);
      storage.setItem(key, nextValue);
    }
  );

  return derivedAtom;
}
```

**Reply by dai-shi** [reply]
See also my version there: https://github.com/pmndrs/jotai/discussions/1879#discussioncomment-5626120

**Reply by chris-cotech** [reply]
I am receiving the following error during `npm run build` in my NextJS app using app router: 

ReferenceError: localStorage is not defined

If there is anything you could please suggest for this. Many thanks!

**Reply by dai-shi** [reply]
How does your code look like?

**Reply by mahdiElmi** [reply]
I get the same "ReferenceError: localStorage is not defined" error like chris.

```ts import { LocalShoppingCartItems } from "@/lib/types";
import { atomWithStorage } from "jotai/utils";

export const cartAtom = atomWithStorage<LocalShoppingCartItems>(
  "shoppingCart",
  localStorage.getItem("shoppingCart") !== null
    ? JSON.parse(localStorage.getItem("shoppingCart")!)
    : {},
);
```
I'm using next js app directory. if I add "use client" I get this error: The "use client" directive must be placed before other expressions. Move it to the top of the file to resolve this issue.

if I do a typeof window !== "undefined" check before accessing local storage I get a hydration error.
I also tried to pass  { getOnInit: true } but that also causes another error.
I feel like I'm not using this as I'm supposed to 😄 any recommendations would be appreciated

**Reply by dai-shi** [reply]
> I also tried to pass { getOnInit: true } but that also causes another error.

Try passing `undefined` for the second argument.

**Reply by mahdiElmi** [reply]
did it like so 
```ts 
export const cartAtom = atomWithStorage<LocalShoppingCartItems>(
  "shoppingCart",
  {},
  undefined,
  { getOnInit: true },
);
```
and still got hydration errors :
Error: Text content does not match server-rendered HTML.

Warning: Text content did not match. Server: "0" Client: "1"
(the text is cart items count that is pulled from local storage on client)

maybe I should mention that I'm using the atom with useHydrateAtoms in two other components to merge the shopping cart state from server and client like so:
```ts 
const [items, setItems] = useAtom(cartAtom);
  useHydrateAtoms([
    [cartAtom, mergeCartItems(cartItemsFromServer, items)],
  ]);
  ```
  all I want to do is to have the client version of shopping cart on local storage be merged with the server version on initial load or as soon as possible without any hydration errors.

**Reply by dai-shi** [reply]
> got hydration errors

Sorry, in that case, you shouldn't specify `getOnInit`.

I'm not following the every details, but I guess the solution for your case is `useEffect`.

**Comment by tnlogy** [comment]
Thanks for an interesting library! I found it confusing when trying out atomWithStorage and atomWithHash, that the initialValue is used instead of the stored value in the first render. I wrote my own like this, and then found this thread with similar solutions.

Since my first render called the event `moveend` on my openlayers map, it just overwrite the stored value on every reload of the page.

```javascript
function atomWithStorage(key, initialValue) {
    const _local = atom(localStorage.hasOwnProperty(key) ? localStorage[key] : initialValue);
    return atom(
        (get) => get(_local),
        (get, set, newValue) => {
            localStorage[key] = newValue;
            set(_local, newValue);
        }
    );
}
```

**Reply by dai-shi** [reply]
Nice finding. We have that recipe in the persistence guide too.

**Reply by luvr-dev** [reply]
That works for me until I do a build where I get an error `ReferenceError: localStorage is not defined`

**Comment by XantreDev** [comment]
![image](https://user-images.githubusercontent.com/57757211/221439307-90e73858-f484-4b75-9fde-6050489c7e50.png)

import { atomWithStorage, createJSONStorage } from 'jotai/utils';
import EncryptedStorage from 'react-native-encrypted-storage';

const storage = createJSONStorage(() => EncryptedStorage);

/**
 *
 * @description Suspends before load
 */
export const atomWithEncryptedStorage = <T>(key: string, defaultValue: T) =>
  atomWithStorage<T>(
    key,
    storage
      .getItem(key)
      .then((v) => (typeof v === 'symbol' ? defaultValue : v))
      .catch(() => defaultValue) as any,
    storage as any,
  );

**Reply by XantreDev** [reply]
Suspend atom before get value

**Comment by quolpr** [comment]
Nice solutions about, but I also want to have localstorage changes listen feature (so if locastorage changed in one tab, atoms in other tabs will get known):

```tsx
import { atom, WritableAtom } from "jotai";
import {
  createJSONStorage,
  RESET,
  unstable_NO_STORAGE_VALUE as NO_STORAGE_VALUE,
} from "jotai/utils";

type Unsubscribe = () => void;
type SetStateActionWithReset<Value> =
  | Value
  | typeof RESET
  | ((prev: Value) => Value | typeof RESET);

export function atomWithWebStorage<Value>(
  key: string,
  initialValue: Value
): WritableAtom<Value, [SetStateActionWithReset<Value>], void> {
  const storage = createJSONStorage(() => window.localStorage);
  const valueFromStorage = storage.getItem(key);
  const baseAtom = atom<Value>(
    valueFromStorage === NO_STORAGE_VALUE ? initialValue : valueFromStorage
  );

  if (import.meta.env.MODE !== "production") {
    baseAtom.debugPrivate = true;
  }

  baseAtom.onMount = (setAtom) => {
    const value = storage.getItem(key);
    setAtom(value === NO_STORAGE_VALUE ? initialValue : value);
    let unsub: Unsubscribe | undefined;
    if (storage.subscribe) {
      unsub = storage.subscribe(key, setAtom);
    }
    return unsub;
  };

  const anAtom = atom(
    (get) => get(baseAtom),
    (get, set, update: SetStateActionWithReset<Value>) => {
      const nextValue =
        typeof update === "function"
          ? (update as (prev: Value) => Value | typeof RESET)(get(baseAtom))
          : update;
      if (nextValue === RESET) {
        set(baseAtom, initialValue);
        return storage.removeItem(key);
      }
      set(baseAtom, nextValue);
      return storage.setItem(key, nextValue);
    }
  );

  return anAtom;
}
```

It is basically an adoption of what jotai already offers(I copy-pasted code), but with this change:

```tsx
const baseAtom = atom<Value>(
  valueFromStorage === NO_STORAGE_VALUE ? initialValue : valueFromStorage
);
```

**Comment by kurdin** [comment]
The above implementation did not work for me, seems like `unstable_NO_STORAGE_VALUE` is not available in the latest version of `jotai/utils`.

Here is working TS version of `atomWithWebStorage` with support of boolean and string as value. 

Using `atomWithWebStorage` instead of  `atomWithStorage` will allow you to persist atom value in the browser's localstorage and restore it current value without `default value` on the first render.

```js
export function atomWithWebStorage<Value>(
	key: string,
	initialValue: Value,
	storage = localStorage
) {
	const storedValue = storage.getItem(key);
	const isString = typeof initialValue === 'string';

	const storageValue = storedValue
		? isString
			? storedValue
			: storedValue === 'true'
		: undefined;

	const baseAtom = atom(storageValue ?? initialValue);

	return atom(
		get => get(baseAtom) as Value,
		(_get, set, nextValue: Value) => {
			set(baseAtom, nextValue);
			storage.setItem(key, nextValue!.toString());
		}
	);
}
```

**Comment by micaiah-effiong** [comment]
@cveering this problem can be resolved by using the `getOnInit` option

```ts
export const userAtom = atomWithStorage('user', null, undefined, { getOnInit: true })
```
**see** https://jotai.org/docs/utilities/storage#parameters

**Reply by VitAndrGuid** [reply]
This is nice, but for SSR / NextJS this will cause a hydration mismatch

**Reply by vietvh-3042** [reply]
@VitAndrGuid I'm having the same problem for the first render. Is there any way to solve it

**Reply by VitAndrGuid** [reply]
@vietvh-3042 not really since the server cannot access local storage the first render cannot have any loaded values.

i ended up with the following monkeypatch:

My root layout is a client component and on it i load the atom values that would intefere with the first render using `useAtomValue` 

On the root layout i have another component with the following code, it basically just waits a little until the localstorage values are loaded and displays a spinner until it does, since the children of this component is rendered after jotai has loaded, the SSR render mismatch wont happen

The problem with this approach is that on the first render of the root layout this spinner will always be shown, that sucks for UX, but if i stayed more hours on this issuei would kms. 

![image](https://github.com/pmndrs/jotai/assets/47070525/d4197e63-1347-4b1a-9adb-20e50aee1174)

see: 
https://www.joshwcomeau.com/react/the-perils-of-rehydration/#abstractions

**Reply by LaurentBonnot** [reply]
> @cveering this problem can be resolved by using the `getOnInit` option
> 
> ```ts
> export const userAtom = atomWithStorage('user', null, undefined, { getOnInit: true })
> ```
> 
> **see** https://jotai.org/docs/utilities/storage#parameters

Nice solution ❤️ However don't forget to update the documentation accordingly
